xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,398 @@
+	}
+	sc, err := gsb.cc.NewSubConn(addrs, opts)
+	if err != nil {
+		return nil, err
+	}
+	gsb.scToSubBalancer[sc] = bal
+	return sc, nil
+}
+
+// close guard for the three functions below?
+func (gsb *gracefulSwitchBalancer) resolveNow(bal balancer.Balancer, opts resolver.ResolveNowOptions) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+
+	// If the resolver specifies an update with a whole new type of policy,
+	// there is no reason to forward a ResolveNow call from the balancer being
+	// switched from, as the most recent update from the Resolver does not
+	// concern the balancer being switched from.
+	if bal == gsb.balancerCurrent && gsb.balancerPending != nil {
+		return
+	}
+	gsb.cc.ResolveNow(opts)
+}
+
+func (gsb *gracefulSwitchBalancer) removeSubConn(bal balancer.Balancer, sc balancer.SubConn) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+	delete(gsb.scToSubBalancer, sc)
+	gsb.cc.RemoveSubConn(sc)
+}
+
+func (gsb *gracefulSwitchBalancer) updateAddresses(bal balancer.Balancer, sc balancer.SubConn, addrs []resolver.Address) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+	gsb.cc.UpdateAddresses(sc, addrs)
+}
+
+// SwitchTo gracefully switches to the new balancer. This function must be
+// called synchronously alongside the rest of the balancer.Balancer methods this
+// Graceful Switch Balancer implements.
+func (gsb *gracefulSwitchBalancer) SwitchTo(builder balancer.Builder) error {
+	if gsb.closed.HasFired() {
+		// logger?
+		return errBalancerClosed
+	}
+	gsb.mu.Lock()
+	ccw := &clientConnWrapper{
+		ClientConn: gsb.cc,
+		gsb:        gsb,
+	}
+	newBalancer := builder.Build(ccw, gsb.bOpts)
+	if newBalancer == nil {
+		// Can this even ever happen?
+		return balancer.ErrBadResolverState
+	}
+	ccw.bal = newBalancer
+	if gsb.balancerCurrent == nil {
+		gsb.balancerCurrent = newBalancer
+		gsb.mu.Unlock()
+		return nil
+	}
+	var balToClose balancer.Balancer
+	gsb.pendingState = balancer.State{
+		ConnectivityState: connectivity.Connecting,
+		Picker:            base.NewErrPicker(balancer.ErrNoSubConnAvailable),
+	}
+	// Clean up resources here that are from a previous pending lb.
+	for sc, sb := range gsb.scToSubBalancer {
+		if sb == gsb.balancerPending {
+			gsb.cc.RemoveSubConn(sc)
+			delete(gsb.scToSubBalancer, sc)
+		}
+	}
+	balToClose = gsb.balancerPending
+	gsb.balancerPending = newBalancer
+	gsb.mu.Unlock()
+	if balToClose != nil {
+		balToClose.Close()
+	}
+	return nil
+}
+
+// UpdateClientConnState() simply forwards the update to one of it's children.
+func (gsb *gracefulSwitchBalancer) UpdateClientConnState(state balancer.ClientConnState) error {
+	if gsb.closed.HasFired() {
+		// logger?
+		return errBalancerClosed
+	}
+	gsb.mu.Lock()
+	balToUpdate := gsb.balancerPending
+	if balToUpdate == nil {
+		balToUpdate = gsb.balancerCurrent
+	}
```

Comment:
Maybe factor into a function?

```go
// caller must hold mu.  (Or this can take it maybe???)
func (gsb...) latestBalancer() balancer.Balancer {
	if gsb.balancerPending != nil {
		return gsb.balancerPending
	}
	return gsb.balancerCurrent
}
```